src/main/scala/org/ergoplatform/nodeView/ErgoNodeViewHolder.scala

```diff
@@ -280,17 +280,17 @@ abstract class ErgoNodeViewHolder[State <: ErgoState[State]](settings: ErgoSetti
       mods.headOption match {
         case Some(h) if h.isInstanceOf[Header] => // modifiers are always of the same type
           val sorted = mods.sortBy(_.asInstanceOf[Header].height)
+          val curHeadersHeight = history().headersHeight
+          val sortedAndHigher = sorted.dropWhile(_.asInstanceOf[Header].height < curHeadersHeight + 1)
 
-          if (sorted.head.asInstanceOf[Header].height == history().headersHeight + 1) {
+          if (sortedAndHigher.nonEmpty) {
             // we apply sorted headers while headers sequence is not broken
-            var expectedHeight = history().headersHeight + 1
             var linkBroken = false
 
-            cfor(0)(_ < sorted.length, _ + 1) { idx =>
-              val header = sorted(idx).asInstanceOf[Header]
-              if (!linkBroken && header.height == expectedHeight) {
+            cfor(0)(_ < sortedAndHigher.length, _ + 1) { idx =>
+              val header = sortedAndHigher(idx).asInstanceOf[Header]
+              if (!linkBroken && header.height == history().headersHeight + 1) {
                 pmodModify(header, local = false)
-                expectedHeight += 1
```

Comment:
@kushti the problem was the silently failing `pmodModify` method, we disregard the failure and increase `expectedHeight` ... so I changed the `expectedHeight` calculation to explicit `history().headersHeight ` call which takes the failure into consideration